storage/interface.go

```diff
@@ -223,6 +223,19 @@ type SelectHints struct {
 	// When disabled, the result may contain samples outside the queried time range but Select() performances
 	// may be improved.
 	DisableTrimming bool
+
+	// Projection hints. They are currently unused in the Prometheus promql engine but can be used by different
+	// implementations of the Queryable interface and engines.
+	// These hints are useful for queries like `sum by (label) (rate(metric[5m]))` - we can safely evaluate it
+	// even if we only fetch the `label` label. For some storage implementations this is beneficial.
+
+	// ProjectionLabels are the minimum amount of labels required to be fetched for this Select call
+	// When honored it is required to add an __series_hash__ label containing the hash of all labels
```

Comment:
Is `__series_hash__` just an example or we want this to be used as a common label for deduplication?

Reply:
Its kinda an example, __series_hash__, we just need some label to distinguis by again that disappears in an aggregation, the name can be chosen to be something that doesnt collide with anything... if someone does "sum by (__series_hash__)" we can use another one

Reply:
Thinking about it - engine should probably tell us which label to add here.. wdyt?

Reply:
Another interesting thing - I think you will need to define how to calculate that series hash because Select() might potentially mix data retrieved from multiple storage layers and they need to have a consistently calculated hash.

Reply:
Should we add a Boolean field asking for storage to add series hash label explicitly?
I don't want to just rely on projection labels to check if storage needs hash label.

Also better to work with disable duplicate check on the engine side

Reply:
> you will need to define how to calculate that series hash because Select() might potentially mix data retrieved from multiple storage layers and they need to have a consistently calculated hash.

That would be responsibility of the Select implementation i think! But we should settle on a particular label.

Reply:
> Should we add a Boolean field asking for storage to add series hash label explicitly? I don't want to just rely on projection labels to check if storage needs hash label. Engine should ask for it
> 
> Also better to work with disable duplicate check on the engine side

Yeah engine needs to disable the duplicate label check when it decides to set hints.

Reply:
But wouldn't it be a problem in Prometheus itself if some blocks had a new format that allowed to take advantage of this new field and there were some blocks interleaved?